torch/_functorch/aot_autograd.py

```diff
@@ -1647,6 +1587,106 @@ def remove_dupe_metadata(
         traced_tangents=traced_tangents,
     )
 
+# Given our ViewAndMutation metadata, this fn constructs a new set of metadata,
+# after adding synthetic base arguments to the function.
+# Most of the work in this fn is slogging through all of the metadata corresponding to inputs,
+# and updating it with our synthetic base calling convention.
+#
+# When config.debug_assert is set, we automatically regenerate the metadata
+# and compare it to this output for sanity.
+#
+# In addition to the updated metadata, also return the list of input indices
+# that will need to be updated in the synthetic base epilogue
+def create_synthetic_base_metadata(
+    m: ViewAndMutationMeta,
+    # Maps each outer argument idx to its inner idx (or, if this outer arg is generated from a
+    # synthetic base, you get a tuple of (i, TensorMeta), telling you the base tensor idx, and view metadata)
+    synthetic_base_info: List[Union[int, Tuple[int, torch.Tensor]]],
+    flat_args: List[Any],
+    flat_args_with_synthetic_bases: List[Any],
+) -> Tuple[ViewAndMutationMeta, List[int]]:
+
+    # For each arg in the inner calling convention, map it to its idx in the outer calling convention
+    # (there might be multiple indices, if the inner arg is a synthetic base)
+    synthetic_base_to_indices: List[List[int]] = []
+    for idx in range(len(flat_args_with_synthetic_bases)):
+        outer_aliased_indices_of_current_base_arg = [
+            i for i, outer_idx_or_tuple in enumerate(synthetic_base_info)
+            if (isinstance(outer_idx_or_tuple, int) and outer_idx_or_tuple == idx)
+            or (isinstance(outer_idx_or_tuple, tuple) and outer_idx_or_tuple[0] == idx)
+        ]
+        synthetic_base_to_indices.append(outer_aliased_indices_of_current_base_arg)
+
+    # given the requires_grad info on mutated inputs,
+    # generate the requires_grad info on those same mutated inputs, but after constructing synthetic bases.
+    input_infos = []
+    mutated_inp_require_grad_info = []
+    for i, indices in enumerate(synthetic_base_to_indices):
+        inpt_info = InputAliasInfo(
+            # If len(indices) > 1, then this input is a synthetic base.
+            # The invariant is that to the rest of aot autograd, synthetic bases only show up if
+            # one of their aliases gets a data mutation. And if any of their aliases get metadata
+            # mutations, they will be hidden from the rest of aot autograd.
+            mutates_data=True if len(indices) > 1 else m.input_info[indices[0]].mutates_data,
+            mutates_metadata=False if len(indices) > 1 else m.input_info[indices[0]].mutates_metadata,
+            # leaf-ness should be all-or-nothing for aliased tensor.
```

Comment:
so what, some sort of assert that all(m.input_info[x].is_leaf) == any(...)?

Reply:
sure, ill add a check